### PR TITLE
Implement dissimilar-origin window.parent and window.top

### LINKED ISSUE #11660: Get window.frameElement to return null
#11644 adds `Window::is_top_level`, which should be used by `Window::GetFrameElement` to return `null` approporiately in a mozbrowser. cc @ConnorGBrewster

### LINKED ISSUE #14996: Cross-origin window.parent
At the moment, a cross-origin `window.parent` returns the `WindowProxy` of the current browsing context, rather than the parent's browsing context. We should always return the parent's `WindowProxy`, it's not until accessing its properties that we should get a security error.

Related to #2382. cc @jdm @Ms2ger @bholley @avadacatavra.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR implements `window.parent` and `window.top` for dissimilar-origin windows.

This PR builds on #15536, only the last commit is part of this PR.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14996 and fix #11660.
- [X] These changes do not require tests because there's already a parentage test in `mozilla/cross-origin-objects/cross-origin-objects.html`.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15799)

### CODE REVIEW

**Comment by asajeffrey** [comment]
cc @jdm @Ms2ger

**Comment by asajeffrey** [comment]
Accidentally fixes #11660.

**Comment by Ms2ger** [comment]
Is there anyone else who could review this?

**Comment by asajeffrey** [comment]
The obvious candidate is @jdm, but he's away next week. @nox?

**Comment by nox** [comment]
I'll as soon as #15536 lands and this gets rebased on top of master.

**Comment by asajeffrey** [comment]
@nox: thanks! If you wanted to take a look at #15536 too, that would be great, it's already gone through quite a bit of review by @jdm and @Ms2ger.

**Comment by asajeffrey** [comment]
@nox: ready when you are!

**Comment by asajeffrey** [comment]
Rebased.

**Comment by nox** [comment]
```
  ▶ TIMEOUT [expected OK] /html/semantics/scripting-1/the-script-element/module/crossorigin.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  └ 3.3 (Core Profile) Mesa 12.0.1

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Root module, Error in CORS-different-origin script

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Root module, Error in CORS-same-origin script

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Root module, Blocked script download, missing CORS ACAO header

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Root module, Blocked script download, mismatched CORS ACAO header

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Imported module, Error in CORS-different-origin script

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Imported module, Error in CORS-same-origin script

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Imported module, Blocked script download, missing CORS ACAO header

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-script-element/module/crossorigin.html:
  └ NOTRUN [expected FAIL] Imported module, Blocked script download, mismatched CORS ACAO header

  ▶ ERROR [expected OK] /workers/data-url.html
  └   → expected expression, got '}'

  ▶ Unexpected subtest result in /workers/data-url.html:
  │ FAIL [expected PASS] worker has opaque origin
  │   → assert_equals: expected "PASS" but got "FAIL"
  │ 
  │ assert_worker_sends_pass/</w.onmessage<@http://web-platform.test:8000/workers/data-url.html:12:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1449:17
```

**Comment by nox** [comment]
```
  ▶ OK [expected TIMEOUT] /html/browsers/browsing-the-web/navigating-across-documents/navigation_unload_data_url.html

  ▶ Unexpected subtest result in /html/browsers/browsing-the-web/navigating-across-documents/navigation_unload_data_url.html:
  └ PASS [expected TIMEOUT] Same-origin navigation started from unload handler

  ▶ OK [expected TIMEOUT] /html/browsers/origin/origin-of-data-document.html

  ▶ Unexpected subtest result in /html/browsers/origin/origin-of-data-document.html:
  │ FAIL [expected TIMEOUT] The origin of a 'data:' document in a frame is opaque.
  │   → assert_equals: Messages sent from a 'data:' URL should have an opaque origin (which serializes to 'null'). expected "null" but got ""
  │ 
  │ @http://web-platform.test:8000/html/browsers/origin/origin-of-data-document.html:14:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1409:20
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1449:17
```

**Comment by nox** [comment]
```
./components/script/script_thread.rs:1707: Line is longer than 120 characters
./components/script/dom/browsingcontext.rs:90: trailing whitespace
./components/script/dom/browsingcontext.rs:111: Line is longer than 120 characters
```

**Comment by asajeffrey** [comment]
Hmm, I didn't to a great job of rebasing this, did I? I'll try to work out why those tests are failing.

**Comment by asajeffrey** [comment]
Ah, `/html/semantics/scripting-1/the-script-element/module/crossorigin.html` is a known intermittent: #15718. The other tests need expectations updated.

File: components/script/script_thread.rs [hunk 1]
```diff
@@ -1650,6 +1650,65 @@ impl ScriptThread {
...
+    }
+
+    fn ask_constellation_for_parent_info(&self, pipeline_id: PipelineId) -> Option<(PipelineId, FrameType)> {
+        let (result_sender, result_receiver) = ipc::channel().unwrap();
+        let msg = ConstellationMsg::GetParentInfo(pipeline_id, result_sender);
+        self.constellation_chan.send(msg).expect("Failed to send to constellation.");
+        result_receiver.recv().expect("Failed to get frame id from constellation.")
+    }
+
+    fn remote_browsing_context(&self,
```
Comment (nox) [inline components/script/script_thread.rs]: What does this do?

File: components/script/script_thread.rs [hunk 2]
```diff
@@ -1650,6 +1650,65 @@ impl ScriptThread {
...
+        let parent = match self.ask_constellation_for_parent_info(pipeline_id) {
+            Some((parent_id, FrameType::IFrame)) => self.remote_browsing_context(global_to_clone, parent_id),
+            _ => None,
+        };
+        let browsing_context = BrowsingContext::new_dissimilar_origin(global_to_clone, frame_id, parent.r());
+        self.browsing_contexts.borrow_mut().insert(frame_id, JS::from_ref(&*browsing_context));
+        Some(browsing_context)
+    }
+
+    fn local_browsing_context(&self,
```
Comment (nox) [inline components/script/script_thread.rs]: What does this do?

File: components/script/script_thread.rs [hunk 3]
```diff
@@ -1650,6 +1650,65 @@ impl ScriptThread {
...
+                              window: &Window,
+                              frame_id: FrameId,
+                              parent_info: Option<(PipelineId, FrameType)>)
+                              -> Root<BrowsingContext>
+    {
+        if let Some(browsing_context) = self.browsing_contexts.borrow().get(&frame_id) {
+            browsing_context.set_currently_active(&*window);
```
Comment (nox) [inline components/script/script_thread.rs]: Why?

File: components/constellation/constellation.rs [hunk 4]
```diff
@@ -1092,7 +1092,14 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             FromScriptMsg::TouchEventProcessed(result) => {
                 self.compositor_proxy.send(ToCompositorMsg::TouchEventProcessed(result))
             }
-
+            FromScriptMsg::GetFrameId(pipeline_id, sender) => {
+                let result = self.pipelines.get(&pipeline_id).map(|pipeline| pipeline.frame_id);
+                let _ = sender.send(result);
```
Comment (nox) [inline components/constellation/constellation.rs]: No `warn!` on send error?

File: components/constellation/constellation.rs [hunk 5]
```diff
@@ -1092,7 +1092,14 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
...
+            }
+            FromScriptMsg::GetParentInfo(pipeline_id, sender) => {
+                let result = self.pipelines.get(&pipeline_id).and_then(|pipeline| pipeline.parent_info);
+                let _ = sender.send(result);
```
Comment (nox) [inline components/constellation/constellation.rs]: No `warn!` on send error?

**Review by nox — COMMENTED** [review]
Some questions.

File: components/constellation/constellation.rs (hunk 4 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Fixed.

File: components/constellation/constellation.rs (hunk 5 above)
Comment (asajeffrey) [inline components/constellation/constellation.rs]: Fixed.

File: components/script/script_thread.rs (hunk 1 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: Gets a browsing context for a pipeline in another script thread. I'll add a comment.

File: components/script/script_thread.rs (hunk 2 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: Gets a browsing context for a pipeline in this script thread. I'll add a comment.

File: components/script/script_thread.rs (hunk 3 above)
Comment (asajeffrey) [inline components/script/script_thread.rs]: The browsing context already exists, but another window is currently active.

**Review by asajeffrey — COMMENTED** [review]
Thanks for the review!

File: components/script/script_thread.rs (hunk 3 above)
Comment (nox) [inline components/script/script_thread.rs]: I don't get it. Why does getting a browsing context for a pipeline in this script thread should change the active window?

**Comment by asajeffrey** [comment]
Rebased.

**Comment by nox** [comment]
Squash it and r=me.

**Comment by asajeffrey** [comment]
Squashed. @bors-servo r=nox

**Pull request merged**